sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/BatchingParDo.java

```diff
@@ -0,0 +1,214 @@
+      this.numElementsInBatchSpec = StateSpecs.value(VarLongCoder.of());
+      this.timerAlreadySetForWindow = StateSpecs.value(VarIntCoder.of());
+      this.keySpec = StateSpecs.value(keyCoder);
+      // prefetch every 20% of batchSize elements. Do not prefetch if batchSize is too little
+      this.prefetchFrequency = ((batchSize / 5) <= 1) ? Long.MAX_VALUE : (batchSize / 5);
+    }
+
+    @ProcessElement
+    public void processElement(
+        @TimerId(END_OF_WINDOW_ID) Timer timer,
+        @StateId(TIMER_ALREADY_SET_ID) ValueState<Integer> timerAlreadySetForWindow,
+        @StateId(BATCH_ID) BagState<InputT> batch,
+        @StateId(NUM_ELEMENTS_IN_BATCH_ID) ValueState<Long> numElementsInBatch,
+        @StateId(KEY_ID) ValueState<K> key,
+        ProcessContext c,
+        BoundedWindow window) {
+      Instant firingInstant = window.maxTimestamp().plus(allowedLateness);
+      // Timers are scoped to the window. A timer can be set only for a single time per scope.
+      // But prevent to set it at each element (set it once per window)
+      Integer isSet = timerAlreadySetForWindow.read();
+      if (isSet == null) {
+        LOGGER.debug(
+            "*** SET TIMER *** to point in time %s for window %s",
+            firingInstant.toString(), window.toString());
+        timer.set(firingInstant);
+        timerAlreadySetForWindow.write(1);
+      }
+      key.write(c.element().getKey());
+      batch.add(c.element().getValue());
+      LOGGER.debug("*** BATCH *** Add element for window %s ", window.toString());
+      Long num = numElementsInBatch.read();
+      if (num == null) {
+        num = 0L;
+      }
+      num++;
+      numElementsInBatch.write(num);
+      if ((num > 0) && (num % prefetchFrequency == 0)) {
+        //prefetch data and modify batch state (readLater() modifies this)
+        batch.readLater();
+      }
+      if (num >= batchSize) {
+        LOGGER.debug("*** END OF BATCH *** for window %s", window.toString());
+        flushBatch(c, key, batch, numElementsInBatch);
+      }
+    }
+
+    @OnTimer(END_OF_WINDOW_ID)
+    public void onTimerCallback(
+        OnTimerContext context,
+        @StateId(KEY_ID) ValueState<K> key,
+        @StateId(BATCH_ID) BagState<InputT> batch,
+        @StateId(NUM_ELEMENTS_IN_BATCH_ID) ValueState<Long> numElementsInBatch,
+        BoundedWindow window) {
+      LOGGER.debug(
+          "*** END OF WINDOW *** for timer timestamp %s in windows %s",
+          context.timestamp(), window.toString());
+      flushBatch(context, key, batch, numElementsInBatch);
+    }
+
+    private void flushBatch(
+        Context c, ValueState<K> key, BagState<InputT> batch, ValueState<Long> numElementsInBatch) {
+      Iterable<OutputT> batchOutput = perBatchFn.apply(batch.read());
+      for (OutputT element : batchOutput) {
+        c.output(KV.of(key.read(), element));
+      }
+      batch.clear();
+      LOGGER.debug("*** BATCH *** clear");
+      numElementsInBatch.write(0L);
```

Comment:
Ideally this would also be `clear()`. If you use `CombiningValueState`, I believe it can be.